In [1]:
import sys
import pandas as pd
print(sys.executable) #check environment. uv from hw1
print(pd.__version__) #check panda version

/Users/kai_97/Desktop/QMSS 5072 HW1/hw01/.venv/bin/python
3.0.6


In [2]:
binge = pd.read_csv("data/binge_prevalence.csv")
pvt = pd.read_csv("data/poverty_prevalence.csv")
binge.head()

,location_abbr,location,year,group,prevalence
0,AK,Alaska,2019,Female,15.2
1,AK,Alaska,2019,Male,10.5
2,AK,Alaska,2019,Overall,13.1
3,AK,Alaska,2021,Female,NaN
4,AK,Alaska,2021,Male,NaN


In [3]:
pvt.head()

,location_abbr,location,year,prevalence
0,AK,Alaska,2019,17.7
1,AK,Alaska,2021,16.7
2,AK,Alaska,2022,17.6
3,AL,Alabama,2019,25.4
4,AL,Alabama,2021,25.8


In [4]:
binge.info()

<class 'pandas.DataFrame'>
RangeIndex: 496 entries, 0 to 495
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   location_abbr  496 non-null    str  
 1   location       496 non-null    str  
 2   year           496 non-null    int64
 3   group          496 non-null    str  
 4   prevalence     367 non-null    str  
dtypes: int64(1), str(4)
memory usage: 19.5 KB


In [5]:
pvt.info()

<class 'pandas.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   location_abbr  156 non-null    str    
 1   location       156 non-null    str    
 2   year           156 non-null    int64  
 3   prevalence     156 non-null    float64
dtypes: float64(1), int64(1), str(2)
memory usage: 5.0 KB


In [6]:
binge_numeric = pd.to_numeric(binge["prevalence"], errors="coerce") #try to turn str into number, keeping unconvertible missing.
binge[
    binge["prevalence"].notna() & binge_numeric.isna()
] # This is intended to check if any value failed conversion.


,location_abbr,location,year,group,prevalence
303,NM,New Mexico,2021,Overall,*


In [7]:
binge["prevalence"]=pd.to_numeric(binge["prevalence"], errors="coerce") #after check, we store converted values to table.
binge.loc[303]

location_abbr            NM
location         New Mexico
year                   2021
group               Overall
prevalence              NaN
Name: 303, dtype: object

 - New Mexico, 2021, Overall: converted the nonnumeric prevalence marker * to a missing value NaN when converting the prevalence column to numeric.

In [8]:
binge[
    binge.duplicated(
        subset=["location_abbr", "year", "group"],
        keep=False
    )
] # check for repeated records. No changes yet. False makes all duplicate occurrences return Ture.

,location_abbr,location,year,group,prevalence
243,MT,Montana,2019,Female,18.3
244,MT,Montana,2019,Female,18.3


In [9]:
binge = binge.drop_duplicates()#By default, check all 5 columns, drop duplicates, keep the first record. Assign back.

 - Montana, 2019, Female: removed one identical duplicate record, keeping one prevalence value of 18.3.

In [10]:
binge[
    binge.duplicated(
        subset=["location_abbr", "year", "group"],
        keep=False
    )
] # check again

,location_abbr,location,year,group,prevalence


In [11]:
pvt[
    pvt.duplicated(
        subset=["location_abbr", "year"],
        keep=False
    )
] #same method check for poverty data.

,location_abbr,location,year,prevalence


In [12]:
binge["location"].unique()

<StringArray>
[              'Alaska',              'Alabama',             'Arkansas',
              'Arizona',           'California',             'Colorado',
          'Connecticut', 'District of Columbia',             'Delaware',
              'Florida',              'Georgia',                 'Guam',
               'Hawaii',                 'Iowa',                'Idaho',
             'Illinois',              'Indiana',               'Kansas',
             'Kentucky',            'Louisiana',        'Massachusetts',
             'Maryland',                'Maine',             'Michigan',
            'Minnesota',             'Missouri',          'Mississippi',
              'Montana',       'North Carolina',         'North Dakota',
             'Nebraska',        'New Hampshire',           'New Jersey',
           'New Mexico',               'Nevada',             'New York',
                 'Ohio',             'Oklahoma',               'Oregon',
         'Pennsylvania',          'Pu

In [13]:
pvt["location"].unique()

<StringArray>
[              'Alaska',              'Alabama',             'Arkansas',
              'Arizona',           'California',             'Colorado',
          'Connecticut', 'District of Columbia',             'Delaware',
              'Florida',              'Georgia',               'Hawaii',
                 'Iowa',                'Idaho',             'Illinois',
              'Indiana',               'Kansas',             'Kentucky',
            'Louisiana',        'Massachusetts',             'Maryland',
                'Maine',             'Michigan',            'Minnesota',
             'Missouri',          'Mississippi',              'Montana',
       'North Carolina',         'North Dakota',             'Nebraska',
        'New Hampshire',           'New Jersey',           'New Mexico',
               'Nevada',             'New York',                 'Ohio',
             'Oklahoma',               'Oregon',         'Pennsylvania',
         'Rhode Island',       'South

In [14]:
binge["location_abbr"].unique()

<StringArray>
['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'GU', 'HI',
 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO',
 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR',
 'PA', 'PR', 'RI', 'SC', 'SD', 'TN', 'TX', 'US', 'UT', 'VA', 'VI', 'VT', 'WA',
 'WI', 'WV', 'WY']
Length: 55, dtype: str

In [15]:
pvt["location_abbr"].unique()

<StringArray>
['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'HI', 'IA',
 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS',
 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA',
 'RI', 'SC', 'SD', 'TN', 'TX', 'US', 'UT', 'VA', 'VT', 'WA', 'WI', 'WV', 'WY']
Length: 52, dtype: str

In [16]:
binge["group"].unique()

<StringArray>
['Female', 'Male', 'Overall']
Length: 3, dtype: str

In [17]:
print("Binge years:", binge["year"].unique())
print("Poverty years:", pvt["year"].unique())

Binge years: [2019 2021 2023]
Poverty years: [2019 2021 2022 2012]


In [18]:
pvt[pvt["year"] ==2012] # Year identify row, outer pvt display row.

,location_abbr,location,year,prevalence
117,RI,Rhode Island,2012,18.5


In [19]:
pvt[pvt["location_abbr"] == "RI"].sort_values("year")

,location_abbr,location,year,prevalence
117,RI,Rhode Island,2012,18.5
118,RI,Rhode Island,2019,16.8
119,RI,Rhode Island,2022,17.0


In [20]:
pvt.loc[
    (pvt["location_abbr"] == "RI") & (pvt["year"] == 2012),
    "year"
] = 2021 #Given the data is from 2019, treat 2012 as typo of 2021, loc and replace it

In [21]:
pvt[pvt["location_abbr"] == "RI"].sort_values("year") #check

,location_abbr,location,year,prevalence
118,RI,Rhode Island,2019,16.8
117,RI,Rhode Island,2021,18.5
119,RI,Rhode Island,2022,17.0


 - Rhode Island: corrected the poverty year from 2012 to 2021.

In [22]:
exclude = ["US", "GU", "PR", "VI"]

binge = binge[~binge["location_abbr"].isin(exclude)].copy()
pvt = pvt[~pvt["location_abbr"].isin(exclude)].copy()

In [23]:
print("Binge rows and columns:", binge.shape)
print("Poverty rows and columns:", pvt.shape)

Binge rows and columns: (459, 5)
Poverty rows and columns: (153, 4)


In [24]:
binge_all = binge.loc[
    binge["group"] == "Overall",
    ["location_abbr", "location", "year", "prevalence"]
].rename(columns={"prevalence": "binge_all"})
#First select rows where group is overall, keep four columns, rename prevalence to bingge_all.
binge_all.head()

,location_abbr,location,year,binge_all
2,AK,Alaska,2019,13.1
5,AK,Alaska,2021,NaN
8,AK,Alaska,2023,9.0
11,AL,Alabama,2019,10.0
14,AL,Alabama,2021,NaN


In [25]:
binge_male = binge.loc[
    binge["group"] == "Male",
    ["location_abbr", "location", "year", "prevalence"]
].rename(columns={"prevalence": "binge_male"})

binge_male.head()

,location_abbr,location,year,binge_male
1,AK,Alaska,2019,10.5
4,AK,Alaska,2021,NaN
7,AK,Alaska,2023,9.7
10,AL,Alabama,2019,9.4
13,AL,Alabama,2021,NaN


In [26]:
binge_female = binge.loc[
    binge["group"] == "Female",
    ["location_abbr", "location", "year", "prevalence"]
].rename(columns={"prevalence": "binge_female"})

binge_female.head()

,location_abbr,location,year,binge_female
0,AK,Alaska,2019,15.2
3,AK,Alaska,2021,NaN
6,AK,Alaska,2023,8.4
9,AL,Alabama,2019,10.5
12,AL,Alabama,2021,NaN


In [27]:
binge_combined = pd.merge(
    binge_all,
    binge_male,
    on=["location_abbr", "location", "year"],
    how="left"
)
#according to keys, merge tables.
binge_combined.head()

,location_abbr,location,year,binge_all,binge_male
0,AK,Alaska,2019,13.1,10.5
1,AK,Alaska,2021,NaN,NaN
2,AK,Alaska,2023,9.0,9.7
3,AL,Alabama,2019,10.0,9.4
4,AL,Alabama,2021,NaN,NaN


In [28]:
binge_wide = pd.merge(
    binge_combined,
    binge_female,
    on=["location_abbr", "location", "year"],
    how="left"
)
#same thing, assign binge_wide since it's side now.
binge_wide.head()

,location_abbr,location,year,binge_all,binge_male,binge_female
0,AK,Alaska,2019,13.1,10.5,15.2
1,AK,Alaska,2021,NaN,NaN,NaN
2,AK,Alaska,2023,9.0,9.7,8.4
3,AL,Alabama,2019,10.0,9.4,10.5
4,AL,Alabama,2021,NaN,NaN,NaN


In [29]:
binge_wide.shape

(153, 6)

In [30]:
pvt_ready = pvt.rename(columns={"prevalence": "poverty"})

In [31]:
state_year = pd.merge(
    binge_wide,
    pvt_ready,
    on=["location_abbr", "location", "year"],
    how="left"
)

state_year.head()

,location_abbr,location,year,binge_all,binge_male,binge_female,poverty
0,AK,Alaska,2019,13.1,10.5,15.2,17.7
1,AK,Alaska,2021,NaN,NaN,NaN,16.7
2,AK,Alaska,2023,9.0,9.7,8.4,NaN
3,AL,Alabama,2019,10.0,9.4,10.5,25.4
4,AL,Alabama,2021,NaN,NaN,NaN,25.8


In [32]:
state_year = state_year.rename(columns={
    "location": "state",
    "location_abbr": "state_abbr"
})
#rename
state_year = state_year[
    ["state", "state_abbr", "year",
     "binge_all", "binge_male", "binge_female", "poverty"]
]
#reorder
state_year.head()

,state,state_abbr,year,binge_all,binge_male,binge_female,poverty
0,Alaska,AK,2019,13.1,10.5,15.2,17.7
1,Alaska,AK,2021,NaN,NaN,NaN,16.7
2,Alaska,AK,2023,9.0,9.7,8.4,NaN
3,Alabama,AL,2019,10.0,9.4,10.5,25.4
4,Alabama,AL,2021,NaN,NaN,NaN,25.8


## Three correction for question 1b
- New Mexico, 2021, Overall: converted the nonnumeric prevalence marker * to a missing value NaN when converting the prevalence column to numeric.
- Montana, 2019, Female: removed one identical duplicate record, keeping one prevalence value of 18.3.
- Rhode Island: corrected the poverty year from 2012 to 2021.

# Question 1c
## Summariz and check

In [33]:
state_year.shape

(153, 7)

In [34]:
state_year.describe()

,year,binge_all,binge_male,binge_female,poverty
count,153.000000,114.000000,115.000000,115.000000,102.000000
mean,2021.000000,11.016667,10.194783,11.720000,20.134314
std,1.638356,3.000941,3.186425,3.164862,4.046646
min,2019.000000,1.000000,1.100000,0.900000,12.000000
25%,2019.000000,9.225000,8.150000,9.850000,16.900000
50%,2021.000000,11.200000,10.200000,11.900000,19.500000
75%,2023.000000,13.100000,12.250000,14.000000,22.275000
max,2023.000000,18.800000,19.000000,18.600000,30.600000


In [35]:
state_year.dtypes

state               str
state_abbr          str
year              int64
binge_all       float64
binge_male      float64
binge_female    float64
poverty         float64
dtype: object

In [36]:
state_year.isna().sum()

state            0
state_abbr       0
year             0
binge_all       39
binge_male      38
binge_female    38
poverty         51
dtype: int64

In [37]:
state_year.to_csv("binge_cleaned.csv", index=False)
#Save a csv file without index

## Question 2: Highest binge drinking in 2023

In [38]:
binge_2023 = state_year[
    state_year["year"] == 2023
].dropna(subset=["binge_all"])
# select 2023, drop missing value
binge_2023.shape

(30, 7)

In [39]:
top10 = binge_2023.sort_values(
    by="binge_all",
    ascending=False
).head(10)
# sort from high to low by overall prevalence, keep the first 10
top10[["state", "binge_male", "binge_female", "binge_all"]]

,state,binge_male,binge_female,binge_all
110,Oklahoma,14.3,13.8,14.2
140,Vermont,12.7,14.3,13.4
80,Montana,12.7,14.0,13.4
8,Arkansas,12.6,12.9,12.8
95,New Jersey,12.3,12.7,12.5
107,Ohio,11.4,12.9,12.2
47,Indiana,12.1,11.7,11.9
92,New Hampshire,10.7,12.6,11.6
20,Connecticut,9.1,14.1,11.6
131,Texas,8.7,14.6,11.5


In [40]:
binge_2023[binge_2023["state_abbr"] == "DC"]

,state,state_abbr,year,binge_all,binge_male,binge_female,poverty
23,District of Columbia,DC,2023,6.8,5.4,8.2,NaN


In [41]:
states_2023 = binge_2023[
    binge_2023["state_abbr"] != "DC"
]
# exclude DC
top10 = states_2023.sort_values(
    by="binge_all",
    ascending=False
).head(10)
# sort again
top10[["state", "binge_male", "binge_female", "binge_all"]]

,state,binge_male,binge_female,binge_all
110,Oklahoma,14.3,13.8,14.2
140,Vermont,12.7,14.3,13.4
80,Montana,12.7,14.0,13.4
8,Arkansas,12.6,12.9,12.8
95,New Jersey,12.3,12.7,12.5
107,Ohio,11.4,12.9,12.2
47,Indiana,12.1,11.7,11.9
20,Connecticut,9.1,14.1,11.6
92,New Hampshire,10.7,12.6,11.6
131,Texas,8.7,14.6,11.5


The ranking is based on 29 states without missing overall high school binge drinking prevalence in 2023. These are the 10 states with the highest binge value, this is not 100% correct since 21 states having missing value.

## Question 3: Poverty and binge drinking

In [42]:
comparison = state_year[
    state_year["year"].isin([2019, 2021])
    & (state_year["state_abbr"] != "DC")
].dropna(subset=["binge_all", "poverty"]).copy()
# select either year (2019 2021), exclude DC, drop missing values (either is missing)
comparison.groupby("year")["state"].count()
#count the eligible states separately for each year

year
2019    42
2021    40
Name: state, dtype: int64

In [43]:
comparison.groupby("year")["poverty"].median()

year
2019    20.0
2021    19.8
Name: poverty, dtype: float64

In [44]:
comparison["poverty_median"] = (
    comparison.groupby("year")["poverty"].transform("median")
)
# give one result per original row
comparison[["state", "year", "poverty", "poverty_median"]].head()

,state,year,poverty,poverty_median
0,Alaska,2019,17.7,20.0
3,Alabama,2019,25.4,20.0
6,Arkansas,2019,27.8,20.0
7,Arkansas,2021,27.2,19.8
9,Arizona,2019,21.9,20.0


In [45]:
comparison["poverty_group"] = "Below median"
# new column poverty_group, all below median
comparison.loc[
    comparison["poverty"] >= comparison["poverty_median"],
    "poverty_group"
] = "At or above median"
# change to at or above after compare
comparison[
    ["state", "year", "poverty", "poverty_median", "poverty_group"]
].head()

,state,year,poverty,poverty_median,poverty_group
0,Alaska,2019,17.7,20.0,Below median
3,Alabama,2019,25.4,20.0,At or above median
6,Arkansas,2019,27.8,20.0,At or above median
7,Arkansas,2021,27.2,19.8,At or above median
9,Arizona,2019,21.9,20.0,At or above median


In [46]:
poverty_summary = (
    comparison
    .groupby(["year", "poverty_group"])["binge_all"]
    .agg(
        number_of_states="count",
        average_binge_rate="mean"
    )
    .reset_index()
)
#create group for year and poverty group, select overall binge level, count and calculate mean, rest index.
poverty_summary

,year,poverty_group,number_of_states,average_binge_rate
0,2019,At or above median,21,12.228571
1,2019,Below median,21,12.352381
2,2021,At or above median,20,11.250000
3,2021,Below median,20,10.645000


## Question 3C
After dropping missing values and excluding DC, the at-or-above-median poverty group had a slightly lower average binge drinking rate in 2019 (12.23%) than the below-median group (12.35%). In 2021, the pattern reversed. The at-or-above-median group drinks a little more (11.25%) than the below-median group (10.65%). Hence, higher drinking was associated with higher poverty only in 2021.